rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2018 RSK Labs Ltd.
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.datasource;
+
+import co.rsk.util.MaxSizeHashMap;
+import org.ethereum.db.ByteArrayWrapper;
+import org.ethereum.util.ByteUtil;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class DataSourceWithCache implements KeyValueDataSource {
+    private final KeyValueDataSource base;
+    private final Map<ByteArrayWrapper, byte[]> uncommittedCache;
+    private final Map<ByteArrayWrapper, byte[]> committedCache;
+    private final Set<ByteArrayWrapper> uncommittedKeysToRemove;
+
+    public DataSourceWithCache(KeyValueDataSource base, int cacheSize) {
+        this.base = base;
+        this.uncommittedCache = new HashMap<>();
+        this.committedCache = new MaxSizeHashMap<>(cacheSize, true);
+        this.uncommittedKeysToRemove = new HashSet<>();
+    }
+
+    @Override
+    public byte[] get(byte[] key) {
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+
+        if (uncommittedKeysToRemove.contains(wrappedKey)) {
+            return null;
+        }
+
+        if (committedCache.containsKey(wrappedKey)) {
+            return committedCache.get(wrappedKey);
+        }
+        if (uncommittedCache.containsKey(wrappedKey)) {
+            return uncommittedCache.get(wrappedKey);
+        }
+
+        byte[] value = base.get(key);
+        committedCache.put(wrappedKey, value);
```

Comment:
null value, as expected, is allowed here to be stored in commitedCache

Reply:
added as a comment